### DISCUSSION TITLE
Creating a custom device that works with Zigbee2mqtt?

### DISCUSSION THREAD

**Discussion by proffalken** [Koenkk/zigbee2mqtt#10461] [Devices]
Hi all,

I'm looking at porting a prototype that I've currently got running via an ESP32 to a zigbee connection rather than a WiFi one.

I'd like to make sure that it's compatible with Zigbee2Mqtt from the outset.

Looking around, the only dev boards I can find are the [XBee boards](https://www.hobbytronics.co.uk/zigbee/xbee-3mw-pcb), however the only XBee device listed on the `support devices` page is a router rather than an end device.

Can anyone recommend a dev board that I can use with Zigbee2mqtt?

**Comment by Ultrawipf** [comment]
I am also very interested in that and would love to develop custom zigbee end devices but the lack of open hardware and software or any examples at all is annoying.

Problem with the CC2530 is the complicated development setup. Something more modern would be great.

I was able to get an nrf52840-dongle devkit to show up in zigbee2mqtt at least and that can be easily programmed with an integrated bootloader via usb and the sdk has a good vscode integration. But haven't gone further yet.

A CC2530 based ebyte E18 also shows up but the options what to do with it are pretty limited with the provided demo firmware.

This project also looks quite interesting: https://github.com/Wiz-IO/wizio-cc

**Reply by stevenpetryk** [reply]
By any chance, would you be willing to share the code that you used to get the nRF52840 to show up in zigbee2mqtt? I'm trying to make a custom end device and would love to have a better starting point than Nordic's super complex example code.

**Reply by Ultrawipf** [reply]
I gave up on that for now but i think it was based on the lightswitch/lightbulb demo from nordic: https://developer.nordicsemi.com/nRF_Connect_SDK_dev/doc/PR-4004/nrf/samples/zigbee/light_bulb/README.html

The code is a bit complicated but it did show up as a zigbee device and might have worked with some more debugging on the zigbee2mqtt side.

**Comment by sjorge** [comment]
You probably want to use a unique zigbeeModule and then write a custom convertor for it: https://www.zigbee2mqtt.io/advanced/support-new-devices/01_support_new_devices.html#_2-adding-your-device

If you plan to sell the hardware in large volumes later you can probably also get the definition merged into zhc.
